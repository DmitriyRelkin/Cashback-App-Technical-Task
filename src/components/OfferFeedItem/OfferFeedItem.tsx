import React from 'react';
import {View, Text, StyleSheet, TouchableOpacity, Image} from 'react-native';
import dayjs from "dayjs";
import {useNavigation} from "@react-navigation/native";
import {useSelector} from "react-redux";

export interface IOfferFeedItem {
  retailerLogo: string;
  title: string;
  expirationDate: string;
  description: string;
  cashbackAmount: number;
}

const OfferFeedItem: React.FC = ({ item }: {item: IOfferFeedItem}) => {
  const navigation = useNavigation();
  const { retailerLogo, title, expirationDate, description, cashbackAmount, id } = item;
  const { data } = useSelector((state: any) => state.cashback);

  const isClaimed = data.indexOf(id) !== -1;
  return (
    <TouchableOpacity
      style={[styles.card, {backgroundColor: isClaimed ? '#BDFFB1' : 'white'}]}
      onPress={() => navigation.navigate('OfferDetails', { offer: item })}>
      <View style={styles.header}>
        <Image source={{uri: retailerLogo}} style={styles.retailerImage} />
        <Text style={styles.title}>{title}</Text>
        <Text style={styles.expiration}>Expired on: {dayjs(expirationDate).format('MMM D, YYYY')}</Text>
      </View>
      <Text style={styles.description}>{description}</Text>
      {!isClaimed ? (
        <Text style={styles.cashback}>Cashback: {cashbackAmount}%</Text>
      ) : (
        <Text style={styles.cashback}>Cashback Applied!</Text>
      )}

    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  card: {
    padding: 16,
    marginVertical: 8,
    marginHorizontal: 16,
    borderRadius: 8,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
  },
  header: {
    alignItems: 'center',
  },
  retailerImage: {
    width: 40,
    height: 40,
    borderRadius: 20,
    marginRight: 8,
  },
  title: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  expiration: {
    fontSize: 14,
    fontStyle: 'italic',
    color: '#810000',
    fontWeight: "700"
  },
  description: {
    fontSize: 16,
    marginVertical: 8,
  },
  cashback: {
    fontSize: 16,
    color: '#099d35',
    fontWeight: "700"
  },
});

export default OfferFeedItem;
