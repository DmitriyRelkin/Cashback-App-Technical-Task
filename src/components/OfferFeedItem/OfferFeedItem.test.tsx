import React from 'react';
import renderer from 'react-test-renderer';
import { useNavigation } from '@react-navigation/native';
import { useSelector } from 'react-redux';
import OfferFeedItem, {IOfferFeedItem} from "./OfferFeedItem";

// Mock the necessary hooks
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

describe('OfferFeedItem', () => {
  const mockNavigation = {
    navigate: jest.fn(),
  };

  const mockUseSelector = useSelector as jest.Mock;
  const mockUseNavigation = useNavigation as jest.Mock;

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseNavigation.mockReturnValue(mockNavigation);
    mockUseSelector.mockReturnValue({ data: [] });
  });

  const mockItem: IOfferFeedItem = {
    retailerLogo: 'https://example.com/logo.png',
    title: 'Sample Offer',
    expirationDate: '2024-12-31T23:59:59.999Z',
    description: 'This is a sample offer description.',
    cashbackAmount: 10,
    id: 1,
  };

  it('renders the component correctly', () => {
    const tree = renderer.create(<OfferFeedItem item={mockItem} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows "Cashback Applied!" when the offer is claimed', () => {
    mockUseSelector.mockReturnValue({ data: [1] });
    const tree = renderer.create(<OfferFeedItem item={mockItem} />);
    const cashbackText = tree.root.findAllByType('Text').find(
      (node) => node.children[0] === 'Cashback Applied!'
    );

    expect(cashbackText).toBeTruthy();
  });
});
