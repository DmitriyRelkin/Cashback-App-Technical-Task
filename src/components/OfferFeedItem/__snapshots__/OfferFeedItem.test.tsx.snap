// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferFeedItem renders the component correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 8,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "opacity": 1,
      "padding": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 8,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://example.com/logo.png",
        }
      }
      style={
        {
          "borderRadius": 20,
          "height": 40,
          "marginRight": 8,
          "width": 40,
        }
      }
    />
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Sample Offer
    </Text>
    <Text
      style={
        {
          "color": "#810000",
          "fontSize": 14,
          "fontStyle": "italic",
          "fontWeight": "700",
        }
      }
    >
      Expired on: 
      Jan 1, 2025
    </Text>
  </View>
  <Text
    style={
      {
        "fontSize": 16,
        "marginVertical": 8,
      }
    }
  >
    This is a sample offer description.
  </Text>
  <Text
    style={
      {
        "color": "#099d35",
        "fontSize": 16,
        "fontWeight": "700",
      }
    }
  >
    Cashback: 
    10
    %
  </Text>
</View>
`;
